Content.Server/Construction/PartExchangerSystem.cs

```diff
@@ -35,35 +36,89 @@ private void OnDoAfter(EntityUid uid, PartExchangerComponent component, DoAfterE
         if (args.Cancelled || args.Handled || args.Args.Target == null)
             return;
 
-        if (!TryComp<MachineComponent>(args.Args.Target.Value, out var machine))
-            return;
-
         if (!TryComp<ServerStorageComponent>(uid, out var storage) || storage.Storage == null)
             return; //the parts are stored in here
 
+        var machinePartQuery = GetEntityQuery<MachinePartComponent>();
+        var machineParts = new List<MachinePartComponent>();
+
+        foreach (var item in storage.Storage.ContainedEntities) //get parts in RPED
+        {
+            if (machinePartQuery.TryGetComponent(item, out var part))
+                machineParts.Add(part);
+        }
+
+        TryExchangeMachineParts(args.Args.Target.Value, storage, machineParts);
+        TryConstructMachineParts(args.Args.Target.Value, storage, machineParts);
+
+        args.Handled = true;
+    }
+
+    private void TryExchangeMachineParts(EntityUid uid, ServerStorageComponent storage, List<MachinePartComponent> machineParts)
+    {
+        if (!TryComp<MachineComponent>(uid, out var machine) || storage.Storage == null)
+            return;
```

Comment:
this would more typically be written by passing in the ServerStorageComponent as nullable with a default value of null, then calling `Resolve(uid, ref storage)`.

Reply:
so:
```
    private void TryExchangeMachineParts(EntityUid uid, List<MachinePartComponent> machineParts,  ServerStorageComponent? storage = null, MachineComponent? machine = null)
    {
        if (!Resolve(uid, ref storage, ref machine))
            return;
```

Reply:
This causes a dereference of a possibly null reference error on line 94 but it would work if I did:
```
private void TryExchangeMachineParts(EntityUid uid, List<MachinePartComponent> machineParts, ServerStorageComponent? storage = null, MachineComponent? machine = null)
    {
        if (!Resolve(uid, ref storage, ref machine) || storage.Storage == null)
            return;
``` 

I just don't know if this is proper coding conventions.

Reply:
After testing it, I found that this code just causes the exchanging to not work at all. So I guess I'll keep it as is.